src/cart/crt_swim.c

```diff
@@ -733,7 +735,25 @@ static void crt_swim_new_incarnation(struct swim_context *ctx,
 	state->sms_incarnation = incarnation;
 }
 
-static int64_t crt_swim_progress_cb(crt_context_t crt_ctx, int64_t timeout, void *arg)
+static void crt_swim_update_last_unpack_hlc(struct crt_swim_membs *csm)
+{
+	struct crt_context	*ctx = NULL;
+	d_list_t		*ctx_list;
+
```

Comment:
even read_lock has its own overhead, because most read_lock is implemented like this: 
```
spin_lock(read_lock->lock);
read_lock->readers++;
spin_unlock(read_lock->lock);
```
Because time granularity is millisecond, so I'd suggest to have:
```
if (hlc - crt_gdata.cg_last_unpack_hlc > 1milliscond) {
        D_RWLOCK_RDLOCK(...);
        ....
        D_RWLOCK_UNLOCK();
        crt_gdata.cg_last_unpack_hlc = hlc; // no lock required for int64
}

Reply:
This lock guard ctx_list from changes during the loop. Your additional condition don't helps because of this function called after swim_progress for each progress call. So, I expect this condition will be always true.

Reply:
Also, "csm->csm_last_unpack_hlc = hlc;" updated only in this place. So, it stay unchanged without this loop.

Reply:
I'm actually suggesting to add a new member of crt_gdata (not  and store the timestamp of the last update, so we don't have to take the lock for every crt_progress(), instead it takes the loop one time per millisecond

Reply:
Hmm. Let's I think more about this. Maybe I will come up with other proposal later (in next patch).